src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceDescriptionService.java

```diff
@@ -498,7 +528,178 @@ public ServiceDescriptionType refreshServiceDescription(Long id, boolean ignoreW
+
+        verifyAuthority("EDIT_OPENAPI3");
+        ServiceDescriptionType serviceDescription = getServiceDescriptiontype(id);
+
+        if (newRestServiceCode == null) {
+            newRestServiceCode = restServiceCode;
+        }
+
+        updateServiceCodes(restServiceCode, newRestServiceCode, serviceDescription);
+
+        // Parse openapi definition and handle updating endpoints and acls
+        if (serviceDescription.getUrl() != url) {
+            parseOpenapi3ToServiceDescription(url, newRestServiceCode, ignoreWarnings, serviceDescription);
+        }
+
+        checkDuplicateServiceCodes(serviceDescription);
+        checkDuplicateUrl(serviceDescription);
+
+        clientRepository.saveOrUpdateAndFlush(serviceDescription.getClient());
+
+        return serviceDescription;
+    }
+
+    /**
+     * Parse OpenApi3 description and update endpoints and acls in ServiceDescription accordingly
+     *
+     * @param url
+     * @param serviceCode
+     * @param ignoreWarnings
+     * @param serviceDescription
+     * @throws OpenApiParser.ParsingException
+     * @throws UnhandledWarningsException
+     */
+    private void parseOpenapi3ToServiceDescription(String url, String serviceCode, boolean ignoreWarnings,
+                                                   ServiceDescriptionType serviceDescription) throws
+            OpenApiParser.ParsingException, UnhandledWarningsException {
+        OpenApiParser.Result result = openApiParser.parse(url);
+        if (!ignoreWarnings && result.hasWarnings()) {
+            WarningDeviation openapiParserWarnings = new WarningDeviation("OpenapiParserWarnings",
+                    result.getWarnings());
+            throw new UnhandledWarningsException(Arrays.asList(openapiParserWarnings));
+        }
+
+        // Update url
+        updateServiceDescriptionUrl(serviceDescription, serviceCode, url);
+
+        // Create endpoints from parsed results
+        List<EndpointType> parsedEndpoints = result.getOperations().stream()
+                .map(operation -> new EndpointType(serviceCode, operation.getMethod(), operation.getPath(),
+                        true))
+                .collect(Collectors.toList());
+        parsedEndpoints.add(new EndpointType(serviceCode, "*", "**", true));
+
+        // Change existing, manually added, endpoints to generated if they're found from parsedEndpoints
+        serviceDescription.getClient().getEndpoint().forEach(ep -> {
+            if (parsedEndpoints.stream().anyMatch(parsedEp -> parsedEp.isEquivalent(ep))) {
+                ep.setGenerated(true);
+            }
+        });
+
+        // Remove ACLs that don't exist in the parsed endpoints list
+        serviceDescription.getClient().getAcl().removeIf(acl ->
+                acl.getEndpoint().isGenerated()
+                        && !parsedEndpoints.stream().anyMatch(endpoint -> acl.getEndpoint().isEquivalent(endpoint)));
```

Comment:
There is also `noneMatch` for a negated `anyMatch`

Reply:
fixed